src/reconcile/rec_write.c

```diff
@@ -567,11 +567,16 @@ __rec_init(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags, WT_SALVAGE_COO
     WT_ORDERED_READ(r->last_running, txn_global->last_running);
 
     /*
-     * Cache the pinned timestamp and oldest id, these are used to when we clear obsolete timestamps
+     * Cache the pinned timestamp and oldest id, these are used when we clear obsolete timestamps
      * and ids from time windows later in reconciliation.
+     *
+     * If we're recovering, all the transaction IDs we see will be from the previous session. Don't
+     * compare them to the current oldest id, which is incommensurate; instead use WT_TXN_MAX so
+     * they will all compare as obsolete.
      */
     __wt_txn_pinned_timestamp(session, &r->rec_start_pinned_ts);
-    r->rec_start_oldest_id = __wt_txn_oldest_id(session);
+    r->rec_start_oldest_id =
+      F_ISSET(S2C(session), WT_CONN_RECOVERING) ? WT_TXN_MAX : __wt_txn_oldest_id(session);
```

Comment:
Setting the WT_TXN_MAX as oldest_id during recovery can lead to a scenario where it clears all the transaction ids that are present on the page before rollback to stable fixes them according to the recovered checkpoint snapshot. The above-described problem can occur mainly for the history store table as this table is accessed for many data tables. Due to this reason, we don't allow the oldest_id to go beyond the recovered checkpoint snapshot_min. Inside the `__wt_txn_oldest_id` function we do the math.

Reply:
Ok, I see how that could be a problem (update a history store page during RTS, it gets evicted, this discards the transaction ids before RTS gets to the history store pass)... and writing out a few stale timestamps isn't really a problem... but comparing transaction ids of different generations still makes me nervous because there are so many ways for it to go wrong.

I'll think about it and come back on Monday. Enjoy your weekend :-)

Reply:
Ok, I don't think this approach is going to work, but I would still like to clean this up because we should really try to avoid comparing transaction ids of different generations. I'll post in the ticket, since that seems like the right place for discussing approaches.

Reply:
The logic of RTS removing the inconsistent updates written to the checkpoint completely depends on the logic of comparing the saved transaction ids on the page. These saved transaction ids are reset when rollback to stable is finished as part of recovery.